### ISSUE TITLE
[REQ] Simplified way to consume async code anywhere with new methods on Task class

### ISSUE THREAD

**Issue by koddek** [dotnet/runtime#83113]
## Background and Motivation
Can this be done?

Now that source generation is available in .NET, it would be beneficial to introduce a new async-await syntax that addresses the existing frustration with async-await when waiting for a result in an area where async-await cannot be used, such as in a property, constructor, or overridden synchronous method. It's important to recognise that async-await cannot be used everywhere, and using workarounds to make async code function properly can become cumbersome. A potential solution is to add methods on the Task class that can wait for results properly and work in any context. The implementation could be source generated to adapt to different usage scenarios.

Note that 'AsyncAwait' and 'AsyncAwaitVoid' are just examples of possible names for these methods. The key idea is to create a new syntax that provides a simpler and more elegant solution for waiting on asynchronous results in contexts where async-await cannot be used directly. This would solve most if not all headaches that comes with consuming async code. And I think if this is implemented properly, it would be heavily used going forward.

## Proposed API

```cs
public class Task
{
    //...
    // AsyncAwait and needed overloads. Could do some source generation to generate needed code
    static T AsyncAwait<T>( /* whatever params (Func<T>) needed to execute Task,Task<T>,ValueTask,ValueTask<T> */);

    // AsyncAwaitVoid and needed overloads. Could do some source generation to generate needed code
    static void AsyncAwaitVoid(/* whatever params (Func<T>) needed to execute Task,Task<T>,ValueTask,ValueTask<T> */);
    //...
}

```

## Usage Examples

``` cs
public string Read(string file)
{
    // can be used in synchronous methods
    // 'AsyncAwait' (or any other suitable name) should wait for the result like async-await, with proper exception handling, and without returning a task. It should return the result directly
    var text = Task.AsyncAwait(() => File.ReadAllTextAsync(file));
    return text.ToUpper();
}

public void Write(string file)
{
    // 'AsyncAwaitVoid' (or any other suitable name) should return void. having Void helps with showing the intent of the code
    // can be used in a void method
    Task.AsyncAwaitVoid(() => File.WriteAllTextAsync(file, "Hello, World!"));
}

public class MyClass
{
    private readonly string text;

    public MyClass()
    {
        // can be used in constructors
        text = Task.AsyncAwait(()=>File.ReadAllTextAsync("file.txt"));
        Task.AsyncAwaitVoid(()=>DoSetupStuffAsync());
    }

    // can be used in properties
    public int Name
    {
        get => Task.AsyncAwait(()=>LocalStorage.GetItemAsync("name"));
        set => Task.AsyncAwait(()=>LocalStorage.SetItemAsync("name", value));
    }

    //...
}
```

## Risks
I don't see any risks so far since this would be new methods on the Task Class.

**Comment by CyrusNajmabadi** [comment]
Not sure how this is different from the existing `.Wait` or `.Result` members on Task.  Could you clarify?

**Comment by koddek** [comment]
.Wait and .Result blocks the thread. I am asking to make these methods work like async-await but just in a method format, where exceptions bubble up properly and are not wrapped, and your app don't crash on errors, and could work in places async await cannot.

**Comment by stephentoub** [comment]
> I am asking to make these methods work like async-await but just in a method format

If the call to AsyncAwait in your example:
```C#
public string Read(string file)
{
    // 'AsyncAwait' (or any other suitable name) should wait for the result like async-await, with proper exception handling, and without returning a task. It should return the result directly
    var text = Task.AsyncAwait(() => File.ReadAllTextAsync(file));
    return text.ToUpper();
}
```
doesn't block, how does `Read` return the string result to its caller synchronously?

**Comment by CyrusNajmabadi** [comment]
> where exceptions bubble up properly and are not wrapped,

How would the exception bubble up... unless you block the current thread waiting on the task to complete and produce it's faulted exception?

**Comment by teo-tsirpanis** [comment]
> I am asking to make these methods work like async-await

Unfortunately if you want to use asynchrony you have to make your call chain asynchronous all the way up. [This is a known problem of asynchronous methods](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/) with no present solution. There is an experiment about adding Green Threads to .NET, where synchronously waiting on a task would `await` it under the hood, but there is no guarantee when or if it would be productized.

> could work in places async await cannot

If you can't use `async`/`await` in some place you can, in descending order of preference:

* Adjust that place so that you can use `async`/`await`; for example if you have a constructor consider a factory method instead.
* If you are consuming 3rd-party code that you can't change, try using a synchronous alternative if exists; performance will be better than the next option.
* Synchronously wait the task using `GetAwaiter.GetResult()` (this does not wrap the exception); performance will be negatively impacted, you should avoid it.

**Comment by koddek** [comment]
> > I am asking to make these methods work like async-await but just in a method format
> 
> If the call to AsyncAwait in your example:
> 
> ```cs
> public string Read(string file)
> {
>     // 'AsyncAwait' (or any other suitable name) should wait for the result like async-await, with proper exception handling, and without returning a task. It should return the result directly
>     var text = Task.AsyncAwait(() => File.ReadAllTextAsync(file));
>     return text.ToUpper();
> }
> ```
> 
> doesn't block, how does `Read` return the string result to its caller synchronously?

- This is a question for the language designers. Is it possible to encapsulate the functionality in a single method like this? It don't have to be all, just the consuming functionality.

> > where exceptions bubble up properly and are not wrapped,
> 
> How would the exception bubble up... unless you block the current thread waiting on the task to complete and produce it's faulted exception?

- My very novice thought process is... inside Task.AsyncAwait, the implementation could be, run the given task in a background thread, wait for the results if any, then return it back to the calling thread. All this would happen automatically in the background. It may require state machines to work like with async await.

**Comment by CyrusNajmabadi** [comment]
> My very novice thought process is... inside Task.AsyncAwait, the implementation could be, run the given task in a background thread, wait for the results if any, then return it back to the calling thread. 

Here's the code that does just htat.  No need for any new additions to the BCL:

```c#
var result = Task.Run(() => YourAsyncCode()).GetAwaiter().GetResult();
```

If you want, you can trivially hide this behind an extension method.  But the above is effectly exactly what you're asking for.  It runs the task in a background thread.  Waits for the results.  And returns it back to the calling thread.

**Comment by koddek** [comment]
> > My very novice thought process is... inside Task.AsyncAwait, the implementation could be, run the given task in a background thread, wait for the results if any, then return it back to the calling thread.
> 
> Here's the code that does just htat. No need for any new additions to the BCL:
> 
> ```cs
> var result = Task.Run(() => YourAsyncCode()).GetAwaiter().GetResult();
> ```
> 
> If you want, you can trivially hide this behind an extension method. But the above is effectly exactly what you're asking for. It runs the task in a background thread. Waits for the results. And returns it back to the calling thread.

I did not know '.GetAwaiter().GetResult()' worked that way as you keep hearing to not use it because it block the thread. Interesting. Thanks. Can you point me to a link that does a deep dive into async await?

**Comment by CyrusNajmabadi** [comment]
>  because it block the thread.

It does block the thread. But those are the semantics you asked for. You asked to:

> wait for the results if any, then return it back to the calling thread.

So you're blocking.  If you want to be non-blocking... Well, you'll have to push that upwards so the caller can then decide what to do.

**Comment by koddek** [comment]
> > because it block the thread.
> 
> It does block the thread. But those are the semantics you asked for. You asked to:
> 
> > wait for the results if any, then return it back to the calling thread.
> 
> So you're blocking. If you want to be non-blocking... Well, you'll have to push that upwards so the caller can then decide what to do.

I am now confused. My original request was for a 'method' that behaves like async-await and can be used anywhere. I thought that using `var result = Task.Run(() => YourAsyncCode()).GetAwaiter().GetResult();` would achieve this, but you are now telling me that it does not.

My point is that there is too much confusion online about how this works. We need a simpler syntax that works universally.

**Comment by theodorzoulias** [comment]
@koddek if I understand correctly, you want the benefits of `async`/`await` without the `async`/`await` itself. It's like saying: *"hey, I like this `async`/`await` thing because it makes my applications so much more responsive and scalable, so please could you do some magic and eliminate it from my code, so that I can reap the benefits of asynchronous programming without having to deal with its complexity?"*. Spoiler alert, what your are asking is not possible, otherwise Microsoft would have already done it. So, unless you are able to explain the mechanism that will make the magic happen, including all the technical and gory details, don't expect that your request will be satisfied any time soon!

**Comment by CyrusNajmabadi** [comment]
>  thought that using var result = Task.Run(() => YourAsyncCode()).GetAwaiter().GetResult(); would achieve this, but you are now telling me that it does not.

It does what you asked for :)  

It's just you're asking for two incompatible things in different forks of this thread.  In one you want a way to use async/await from a sync method, waiting until the results is available (that's what hte above does).  However, in another, you say you do not want to block.  These requests are incompatible with each other.  "waiting until the result is available" is "blocking".  You have to either choose if you want that, or if you want to be non-blocking.

**Comment by koddek** [comment]
> r are asking is not possible

Ok I understand this. Thanks for explaining.

**Comment by koddek** [comment]
> > thought that using var result = Task.Run(() => YourAsyncCode()).GetAwaiter().GetResult(); would achieve this, but you are now telling me that it does not.
> 
> It does what you asked for :)
> 
> It's just you're asking for two incompatible things in different forks of this thread. In one you want a way to use async/await from a sync method, waiting until the results is available (that's what hte above does). However, in another, you say you do not want to block. These requests are incompatible with each other. "waiting until the result is available" is "blocking". You have to either choose if you want that, or if you want to be non-blocking.

Ok I now understand. Thank you for explaining. I guest This request can be closed now.
